actionpack/lib/action_dispatch/testing/integration.rb

```diff
@@ -366,6 +366,11 @@ def remove! # :nodoc:
       def open_session
         dup.tap do |session|
           session.reset!
+
+          parent = self
+          session.define_singleton_method(:assertions) { parent.assertions }
```

Comment:
Instead of dynamically defining the method just define it in the Runner module.

Reply:
🤔 I don't think that would work.

If I defined it on `Runner` wouldn't it apply to all instances of `IntegrationTest`? We need the parent to keep the original implementation, or else we'll have infinite recursion once it hits the parent.

```ruby
class Parent
  attr_accessor :attribute

  def child
    dup.tap do |duplicate|
      parent = self
      child.define_singleton_method(:attribute) { parent.attribute }
      child.define_singleton_method(:attribute=) { |a| parent.attribute = a }
    end
  end
end
```

That's the behaviour I want, in terms of all children delegating to the parent instance for `attribute`. I'm not sure how I'd be able to delegate to the parent without doing it dynamically, since we have no reference to it outside `open_session`.

Am I missing something?

Reply:
I suppose an alternative approach could be something like this.

```ruby
class ActionDispatch::Integration::Session::Runner
  attr_accessor :parent_session

  def assertions
    parent_session ? parent_session.assertions : super
  end

  def assertions=(assertions)
    parent_session ? parent_session.assertions = assertions : super
  end

  def open_session
    dup.tap do |session|
      session.reset!

      self.parent_session = self

      yield session if block_given?
    end
  end
end
```

Moves the dynamic logic into the method, rather than dynamically defining the method itself.

Reply:
Yup, that is what I was referring. Just make sure to add `# :nodoc:` on the attribute and new methods.